// dwdav/index.js
'use strict';

var request = require('request');
var parse = require('xml-parser');
var _ = require('lodash');
var fs = require('fs');
var path = require('path');

function DWDAV (config) {
	this.config = _.extend({
		hostname: 'localhost',
		username: 'admin',
		password: 'password',
		version: 'version1',
		root: '.'
	}, config);
}

DWDAV.prototype.getOpts = function () {
	var opts = {
		baseUrl: 'https://' + this.config.hostname + '/on/demandware.servlet/webdav/Sites/Cartridges/' + this.config.version,
		uri: '/',
		auth: {
			user: this.config.username,
			password: this.config.password
		},
		strictSSL: false
	};
	// Support for 2fa
	if (this.config.p12 && this.config.hostname.indexOf('cert') === 0) {
		opts.strictSSL = true;
		opts.pfx = fs.readFileSync(this.config.p12);
		opts.passphrase = this.config.passphrase;
		opts.honorCipherOrder = true;
		opts.securityOptions = 'SSL_OP_NO_SSLv3';
		opts.secureProtocol = 'TLSv1_1_method';
		// see http://stackoverflow.com/questions/14088787/hostname-ip-doesnt-match-certificates-altname
		// and https://nodejs.org/api/tls.html#tls_tls_connect_port_host_options_callback
		opts.checkServerIdentity = function () {}; 
	}
	return opts;
};

DWDAV.prototype.get = function (filePath, root) {
	var self = this;
	var req;
	var rootFolder = path.join(process.cwd(), root || self.config.root);
	var uriPath = path.relative(rootFolder, filePath || '.');
	var promise = new Promise(function (resolve, reject) {
		req = request(_.extend(self.getOpts(), {
			headers: {
				Depth: 1
			},
			uri: '/' + uriPath,
			method: 'PROPFIND'
		}), function (err, res, body) {
			if (err) {
				return reject(err);
			}
			if (res.statusCode >= 400) {
				return reject(new Error(res.statusMessage));
			}
			var response = parse(body);
			// get "response" children
			var responses = _.filter(response.root.children, function (c) {
				return c.name === 'response';
			})
			// get href and display name of each response
			.map(function (res) {
				var href = _.findWhere(res.children, {name: 'href'}).content;
				var prop = _.findWhere(_.findWhere(res.children, {name: 'propstat'}).children, {name: 'prop'});
				var displayname = _.findWhere(prop.children, {name: 'displayname'}).content;
				return {
					href: href,
					name: displayname
				};
			});
			return resolve(responses);
		});
	});
	promise.request = req;
	return promise;
};

DWDAV.prototype.post = function (filePath, root) {
	var self = this;
	var req;
	var rootFolder = path.join(process.cwd(), root || self.config.root);
	var uriPath = path.relative(rootFolder, filePath);
	var promise = new Promise(function (resolve, reject) {
		req = request(_.extend(self.getOpts(), {
			uri: '/' + uriPath,
			method: 'PUT'
		}), function (err, res, body) {
			if (err) {
				return reject(err);
			}
			if (res.statusCode >= 400) {
				return reject(new Error(res.statusMessage));
			}
			resolve(body);
		});
		fs.createReadStream(filePath).pipe(req);
	});
	promise.request = req;
	return promise;
};

DWDAV.prototype.unzip = function (filePath, root) {
	var self = this;
	var req;
	var rootFolder = path.join(process.cwd(), root || self.config.root);
	var uriPath = path.relative(rootFolder, filePath);

	var promise = new Promise(function (resolve, reject) {
		req = request(_.extend(self.getOpts(), {
			uri: '/' + uriPath,
			method: 'POST',
			form: {
				method: 'UNZIP'
			}
		}), function (err, res, body) {
			if (err) {
				return reject(err);
			}
			if (res.statusCode >= 400) {
				return reject(new Error(res.statusMessage));
			}
			resolve(body);
		});
	});
	promise.request = req;
	return promise;
};

DWDAV.prototype.postAndUnzip = function (filePath, root) {
	var self = this;
	return self.post(filePath, root)
		.then(function () {
			return self.unzip(filePath, root);
		});
};

DWDAV.prototype.delete = function (filePath, root) {
	var self = this;
	var req;
	var rootFolder = path.join(process.cwd(), root || self.config.root);
	var uriPath = path.relative(rootFolder, filePath);

	var promise = new Promise(function (resolve, reject) {
		req = request(_.extend(self.getOpts(), {
			uri: '/' + uriPath,
			method: 'DELETE'
		}), function (err, res, body) {
			if (err) {
				return reject(err);
			}
			// it's ok to ignore 404 error if the file is not found
			if (res.statusCode >= 400 && res.statusCode !== 404) {
				return reject(new Error(res.statusMessage));
			}
			resolve(body);
		});
	});
	promise.request = req;
	return promise;
};

module.exports = function (config) {
	return new DWDAV(config);
};

module.exports.DWDAV = DWDAV;
